README.md
Shuriken Cookbook
=================
[Shuriken](https://github.com/prawn-cake/shuriken) is a monitoring agent which allows to do passive checks for Shinken monitoring system via [mod-ws-arbiter](https://github.com/shinken-monitoring/mod-ws-arbiter).

This cookbook allows to install and manage monitoring agents are installed on dozens of servers easily.

Requirements
------------
* [Python cookbook](https://github.com/poise/python.git)

Attributes
----------
Monitoring agent configuration

- `node[:shuriken][:version]` - shuriken version which will be installed via pip
- `node[:shuriken][:home]` - shuriken home directory, defaule: `/opt/shuriken`
- `node[:shuriken][:user]` - system user, default: `shuriken`
- `node[:shuriken][:group]` - system group, default: `shuriken`
- `node[:shuriken][:logdir]` - shuriken log directory, default: `/var/log/shuriken`
- `node[:shuriken][:plugins_dirs]` - shuriken plugins directories list, default: `["/usr/lib/nagios/plugins"]`. This parameter set a scope of check scipts. It can contains multiple pathes, shuriken will build commands index automatically. If some commands will be have duplicated names in different directories in the scope, then last founded will be used. If no one will be found in the scope, agent will be stopped and raised an Exception. In this case see a log file according to `node[:shuriken][:logdir]`


Monitoring server configuration

```ruby
    node[:shuriken][:server] = {
        :host       => <ip or domain name with schema, for example: http://myshinken.server.com>,
        :port       => <shinken arbiter port>,
        :username   => <username, configured on server side>,
        :password   => <password, configured on server side>,
    }
```

Monitoring check commands configuration
---------------------------------------
Main objective of commands configuration is to provide check commands for agent as json config file.

Examples are provided at `attributes/config.rb`

There is monitored server definition block:

```ruby
    "localhost" => {
        :check_interval => 5,  # in minutes
        :check_commands => {
           "Load"   => "check_load -w2.50,2.20,2.00 -c3.0,2.7,2.5",
           "Disks/" => "check_disk -w 7% -c 5% -p /",
           "Disks/var" => "check_disk -w 7% -c 5% -p /var",
           "Swap"   => "check_swap -a -w15% -c10%"
        }
    },
```

The block above includes several parameters

- `"localhost"` - is a machine hostname FQDN. It's very important parameter allows to identify monitoring configuration. **NOTE:** To setting up hostname in a configuration run `hostname --fqdn` on a monitored machine if don't know it exactly.

- `:check_interval` parameter sets time interval to check execution in minutes. Agent is run by cron.

- `:check_commands` hash

    * **Key** (`"Load"`, `"Disks/"`, etc) is a service description. All services are defined on a monitoring server.
    * **Value** (`"check_load -w2.50,2.20,2.00 -c3.0,2.7,2.5"`, etc) is a raw check command to execute. Shuriken will find absolute path to check command script automatically based on `node[:shuriken][:plugins_dirs]` list. Recipe contains `nagios-plugins` installer therefore you will be able to use nagios plugins by default. 

**NOTE:** Shuriken sanitizes check commands to prevent unsafe command execution.

Usage
-----
- Configure all attribute agent and server parameters;
- Define check commands configurations in `attributes/config.rb` (`"shuriken": {"config": {...}}` in a role)
- Add `'recipe[shuriken]'` to your run_list in a role;
- Add role to a server;

Shinken adapter recipe
-----------------------
Recipe is must used on server side as a last one in a run_list. Recipe parses `node[:shuriken][:config]` as hosts and services and generate new configurations for it and put it onto a server, that means all default and other server attriubutes is read from `node[:shinken]` and other server [cookbook](https://github.com/prawn-cake/shinken-cookbook.git) related files.

Contributing
------------

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
License: MIT

Authors: Maksim Ekimovskiy


CHANGELOG.md
shuriken CHANGELOG
==================

This file is used to list changes made in each version of the shuriken cookbook.

0.2.0
-----
- [prawn-cake] - Added shinken adapter recipe to generate hosts and services configurations by shuriken config


0.1.0
-----
- [prawn-cake] - Initial release of shuriken cookbook

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/default.rb
#
# Cookbook Name:: shuriken
# Recipe:: default
#
# Copyright 2014, Maksim Ekimovskii
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

package 'nagios-plugins'

include_recipe 'python'

# Create home directory
directory "#{node[:shuriken][:home]}" do
    owner   node[:shuriken][:user]
    group   node[:shuriken][:group]
    mode    '0755'
    action  :create
end

# Create system user
user "#{node[:shuriken][:user]}" do
    home    node[:shuriken][:home]
    shell   '/bin/bash'
    action  :create
end

virtualenv_path = "#{node[:shuriken][:home]}"
python_virtualenv virtualenv_path do
    owner   node[:shuriken][:user]
    group   node[:shuriken][:group]
    action  :create
end


# Create log directory
directory "#{node[:shuriken][:logdir]}" do
    owner   node[:shuriken][:user]
    group   node[:shuriken][:group]
    mode    '0755'
    action  :create
end

python_pip "shuriken" do
    virtualenv  virtualenv_path
    version     node[:shuriken][:version]
    user        node[:shuriken][:user]
    group       node[:shuriken][:group]
    action      :install
end

# Setting up shuriken config
require 'json'

shuriken_config_path = "#{node[:shuriken][:home]}/config.json"
template shuriken_config_path do
    source      "config.json.erb"
    owner       node[:shuriken][:user]
    group       node[:shuriken][:group]
    mode        0644
    variables   ({
        "commands" => JSON.pretty_generate(node[:shuriken][:config][node[:fqdn]][:check_commands])
    })
    only_if      { node[:shuriken][:config].has_key?("#{node[:fqdn]}") }
end

PYTHON = "#{node[:shuriken][:home]}/bin/python"
WRAPPER_PATH = "/usr/bin/shuriken"
template WRAPPER_PATH do
    source      "shuriken.erb"
    owner       node[:shuriken][:user]
    group       node[:shuriken][:group]
    mode        0755
    variables   ({
        :PYTHON         => PYTHON,
        :SHURIKEN_HOME  => node[:shuriken][:home]
    })
    #not_if {File.exists?(WRAPPER_PATH)}
end

# Setting up cron task
cron "setup-cron-shuriken" do
    action  :create
    minute  "*/#{node[:shuriken][:config][node[:fqdn]][:check_interval]}"
    user    node[:shuriken][:user]
    command %Q{/usr/bin/shuriken -C "#{shuriken_config_path}" --log "#{node[:shuriken][:logdir]}/shuriken.log"}
    #only_if      { node[:shuriken][:config].has_key?("#{node[:fqdn]}") }
end


recipes/shinken_adapter.rb
# Recipe for generating hosts and services shinken passive configurations based on shuriken configurations


# clean old hosts and services configurations before update
execute "clean-hosts-and-services-conf" do
    command "rm -f #{node[:shinken][:settings]}/hosts/*.cfg && rm -f #{node[:shinken][:settings]}/services/*.cfg"
    user    node[:shinken][:user]
    action  :run
end

# Iterate over shuriken configs and created new hosts and services configurations based on this
node[:shuriken][:config].each do |hostname, params|
    host_conf = {
        :host_name      => hostname,
        :alias          => hostname,
        :address        => hostname,
        :check_command  => '_echo',
        :register       => 1
    }
    
    # TODO: add freshness parameter
    template "/etc/shinken/hosts/#{hostname}.cfg" do
      source 'generic-definition.cfg.erb'
      owner  node['shinken']['user']
      group  node['shinken']['group']
      mode   0644
      variables(
        type: 'host',
        entities: [host_conf]
      )
     # notifies :restart, 'service[shinken-arbiter]'
    end
    
    # TODO: add freshness parameter
    SERVICES = []
    params[:check_commands].each do |svc_name, cmd|
        SERVICES.push({
            :host_name              => hostname,
            :service_description    => svc_name,
            :passive_checks_enabled => 1,
            :check_command          => '_echo',
            :contact_groups         => 'admins',
            :register               => 1
        }.merge(node[:shinken][:service_defaults]))
    end

    template "#{node[:shinken][:settings]}/services/#{hostname}.cfg" do
        source 'generic-definition.cfg.erb'
        owner  node['shinken']['user']
        group  node['shinken']['group']
        mode   0644
        variables(
          type: 'service',
          entities: SERVICES
        )
        notifies :restart, 'service[shinken]'
  end

end


attributes/config.rb
# Global config for all shurikens

# Config format:
# <hostname>: <parameters>
#
# Where parameters are:
# * check_interval -- cron time parameter in minutes;
# * check_commands -- raw monitoring check commands without path to plugins, default plugins path is set in attributes;
# * contact_groups -- set existed contact groups for service checks
#
#
# NOTE: hostname must be equal server fqdn -- check `hostname --fqdn` on the monitoring machine
#
# Check commands:
# <ServiceDescription> => <RawMonitoringShellCommand>
#
#
default[:shuriken][:config] = {
#    "localhost" => {
#        :check_interval => 5,  # in minutes
#        :contact_groups => [], # TODO: implement support of contact groups
#        :check_commands => {
#            "Load"   => "check_load -w2.50,2.20,2.00 -c3.0,2.7,2.5",
#            "Disks/" => "check_disk -w 7% -c 5% -p /",
#            "Disks/var" => "check_disk -w 7% -c 5% -p /var",
#            "Swap"   => "check_swap -a -w15% -c10%"
#        }
#    },
#    "test.local" => {
#        :check_interval => 5,  # in minutes
#        :check_commands => {
#            "Load"   => "check_load -w2.50,2.20,2.00 -c3.0,2.7,2.5",
#            "Disks/" => "check_disk -w 7% -c 5% -p /",
#            "Disks/var" => "check_disk -w 7% -c 5% -p /var",
#            "Swap"   => "check_swap -a -w15% -c10%"
#        }
#    }

}
